web-api/src/mind.rs

```diff
@@ -41,6 +45,39 @@ struct Article {
     url: String,
 }
 
+#[derive(Debug, Deserialize)]
+struct ArticleProvider {
+    articles: HashMap<String, Article>,
+}
+
+impl ArticleProvider {
+    fn new(path: &str) -> Result<Self, anyhow::Error> {
+        let mut articles = HashMap::new();
+        for article in read::<Article>(path)? {
+            let article = article?;
+            articles.insert(article.id.clone(), article);
+        }
+        Ok(Self { articles })
+    }
+
+    fn sample(&self, n: usize) -> Vec<Article> {
+        let mut rng = thread_rng();
+        self.articles
+            .values()
+            .cloned()
+            .collect_vec()
+            .iter()
+            .choose_multiple(&mut rng, n)
+            .into_iter()
+            .cloned()
+            .collect()
+    }
+
+    fn get(&self, id: &str) -> Option<Article> {
+        self.articles.get(id).cloned()
+    }
+}
+
 struct SnippetLabelPair(String, String);
```

Comment:
:art: somewhat related to the `get()` method, we might not even need owned snippets and labels in many contexts. so for example we could operate on references like
```suggestion
struct SnippetLabelPair<'a, 'b>(&'a str, &'b str);
```
this might not fit all our purposes though. if we need to be flexible with owned and referenced types and still want to avoid unnecessary cloning, then feel free to have a look at the `std::borrow::Cow` type:
```suggestion
struct SnippetLabelPair<'a, 'b>(Cow<'a, str>, Cow<'b, str>);
```

no need to change it right away, but it might be a nice exercise for a follow up pr.